estimagic/tests/visualization/test_comparison_plot.py

```diff
@@ -0,0 +1,404 @@
+def test_determine_figure_height_given(df):
+    expected = 400
+    assert _determine_figure_height(df, 400) == expected
+
+
+# _add_plot_specs_to_df
+# ===========================================================================
+
+
+def test_add_plot_specs_to_df():
+    df = pd.DataFrame()
+    df["final_value"] = [0.5, 0.2, 5.5, 4.5, -0.2, -0.1, 4.3, 6.1]
+    df["group"] = list("aabb") * 2
+    df["model_class"] = ["c"] * 4 + ["d"] * 4
+    df["full_name"] = ["a_1", "a_2", "b_1", "b_2"] * 2
+    df["conf_int_lower"] = pd.np.nan
+    df["conf_int_upper"] = pd.np.nan
+    lower, upper, rect_widths = _create_bounds_and_rect_widths(df)
+    color_dict = {}
+
+    expected = df.copy()
+    expected["color"] = "#035096"
+    expected["dodge"] = 0.5
+    expected["lower_edge"] = [0.500, 0.192, 5.488, 4.480, -0.200, -0.102, 4.3, 6.100]
+    expected["upper_edge"] = [0.514, 0.206, 5.524, 4.516, -0.186, -0.088, 4.336, 6.136]
+    expected["binned_x"] = expected[["upper_edge", "lower_edge"]].mean(axis=1)
+
+    _add_plot_specs_to_df(df, rect_widths, lower, upper, color_dict)
+    df[["lower_edge", "upper_edge"]] = df[["lower_edge", "upper_edge"]].round(3)
+
+    assert_frame_equal(df, expected)
+
+
+# _add_color_column
+# ===========================================================================
+
+
+def test_add_color_column_no_dict():
+    color_dict = {}
+    df = pd.DataFrame()
+    df["model_class"] = list("ababab")
+    expected = df.copy(deep=True)
+    expected["color"] = "#035096"
+    _add_color_column(df=df, color_dict=color_dict)
+    assert_frame_equal(df, expected)
+
+
+def test_add_color_column_with_dict():
+    color_dict = {"a": "green"}
+    df = pd.DataFrame()
+    df["model_class"] = list("ababab")
+    expected = df.copy(deep=True)
+    expected["color"] = ["green", "#035096"] * 3
+    _add_color_column(df=df, color_dict=color_dict)
+    assert_frame_equal(df, expected)
+
+
+# _add_dodge_and_binned_x
+# ===========================================================================
+
+
+def test_add_dodge_and_binned_x_without_class():
+    df = pd.DataFrame()
+    df["final_value"] = [1, 5, 3, 3.5, 0.1, 2.5, 0.2, 0.3, 10]
+    df["full_name"] = "param"
+    df["model_class"] = "no class"
+    param = "param"
+    bins = np.array([-2, -1, 0, 1, 2, 3, 4, 5, 6], dtype=float)
+
+    expected = df.copy(deep=True)
+    expected["lower_edge"] = np.array([1, 5, 3, 3, 0, 2, 0, 0, 6], dtype=float)
+    expected["upper_edge"] = np.array([2, 6, 4, 4, 1, 3, 1, 1, np.nan], dtype=float)
+    expected["dodge"] = np.array(
+        [np.nan, np.nan, 0.5, 1.5, 0.5, np.nan, 1.5, 2.5, np.nan], dtype=float
+    )
+    expected["binned_x"] = np.array(
+        [1.5, 5.5, 3.5, 3.5, 0.5, 2.5, 0.5, 0.5, np.nan], dtype=float
+    )
+
+    _add_dodge_and_binned_x(df, param, bins)
+    assert_frame_equal(df, expected)
+
+
+# _find_next_lower
+# ===========================================================================
+
+
+def test_find_next_lower_with_sorted():
```

Comment:
These repetitive patters can be simplified with [``pytest.mark.parametrize``](https://docs.pytest.org/en/latest/parametrize.html).

Reply:
Yes, that makes it much more readable! Switched in a2e17d558f39fcfb95973f9a0559bb1c46f7fe81.